src/ChessBoard/MovePieces.tsx

```diff
@@ -1,134 +1,493 @@
+                            const stepX = validMoves[index].x - currX > 0 ? squareSize : -squareSize;
+                            const stepY = validMoves[index].y - currY > 0 ? squareSize : -squareSize;
+                            let pathX = currX + stepX;
+                            let pathY = currY + stepY;
+                            let blocked = false;
+                            while (pathX !== validMoves[index].x && pathY !== validMoves[index].y) {
+                                if (isPieceOnSquare(pathX, pathY)) {
+                                    blocked = true;
+                                    break;
+                                }
+                                pathX += stepX;
+                                pathY += stepY;
+                            }
+                            isValidMove = !blocked;
+                        }
+                    }
+                    if (isValidMove) break;
+                }
+                if (!isValidMove){ x = currX; y = currY }
+            }
+
+            if (color_name_array[i].name === "king") {
+                //Get all the coordinates where the king can move
+                const left = {x: currX - squareSize, y: currY,
+                    index: getIndexAtPosition(currX - squareSize, currY)}
+
+                const right = {x: currX + squareSize, y: currY,
+                    index: getIndexAtPosition(currX + squareSize, currY)}
+
+                const down = {x: currX, y: currY - squareSize,
+                    index: getIndexAtPosition(currX, currY - squareSize)}
+
+                const up = {x: currX, y: currY + squareSize,
+                    index: getIndexAtPosition(currX, currY + squareSize)}
+
+                const downLeft = {x: currX - squareSize, y: currY - squareSize,
+                    index: getIndexAtPosition(currX - squareSize, currY- squareSize)}
+
+                const downRight = {x: currX + squareSize, y: currY - squareSize,
+                    index: getIndexAtPosition(currX + squareSize, currY - squareSize)}
+
+                const upLeft = {x: currX - squareSize, y: currY + squareSize,
+                    index: getIndexAtPosition(currX - squareSize, currY + squareSize)}
+
+                const upRight = {x: currX + squareSize, y: currY + squareSize,
+                    index: getIndexAtPosition(currX + squareSize, currY + squareSize)}
+
+                const validMoves = [left, right, down, up, downLeft, downRight, upLeft, upRight]
+
+                if (!isValidMove_king_knight(validMoves, x, y, i)) { x = currX; y = currY }
+            }
         }
-        // Only update the state if the mouse has been moved
-        if (x !== this.state.mousePosition.x || y !== this.state.mousePosition.y) {
-            const positions = [...this.state.positions];
-            positions[this.state.draggingIndex] = { x, y };
-            this.setState({ positions, mousePosition: { x, y } });
+        const pos = [...positions];
```

Comment:
this also must be moved to `if (i !== -1) {}` statement. Or you can simply add `if (i === -1) return` at the begining